pkg/ruler/ruler.go

```diff
@@ -750,52 +771,200 @@ func (r *Ruler) getLocalRules(userID string) ([]*GroupStateDesc, error) {
 	return groupDescs, nil
 }
 
-func (r *Ruler) getShardedRules(ctx context.Context, userID string) ([]*GroupStateDesc, error) {
-	ring := ring.ReadRing(r.ring)
+func (r *Ruler) getShardedRules(ctx context.Context, userID string, quorumType QuorumType) ([]*GroupStateDesc, error) {
+	infoLogger := level.Info(r.logger)
+	debugLogger := level.Debug(r.logger)
+
+	type instanceResult struct {
+		res      []*GroupStateDesc
+		err      error
+		instance *ring.InstanceDesc
+	}
+
+	type groupCounter struct {
+		group *GroupStateDesc
+		count int
+	}
+
+	rulerRing := ring.ReadRing(r.ring)
 
 	if shardSize := r.limits.RulerTenantShardSize(userID); shardSize > 0 && r.cfg.ShardingStrategy == util.ShardingStrategyShuffle {
-		ring = r.ring.ShuffleShard(userID, shardSize)
+		rulerRing = r.ring.ShuffleShard(userID, shardSize)
 	}
 
-	rulers, err := ring.GetReplicationSetForOperation(RingOp)
+	rulers, err := rulerRing.GetReplicationSetForOperation(RingOp)
 	if err != nil {
 		return nil, err
 	}
 
+	if rulerRing.ReplicationFactor() <= 2 {
+		// strong quorums are only supported for RF >= 3
+		quorumType = Weak
+	}
+
+	ch := make(chan instanceResult, len(rulers.Instances))
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
 	ctx, err = user.InjectIntoGRPCRequest(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("unable to inject user ID into grpc request, %v", err)
 	}
 
-	var (
-		mergedMx sync.Mutex
-		merged   []*GroupStateDesc
-	)
+	// Spawn a goroutine for each instance.
+	for i := range rulers.Instances {
+		go func(ruler *ring.InstanceDesc) {
+			result, err := r.getDelegateRules(ctx, ruler)
+			ch <- instanceResult{
+				res:      result,
+				err:      err,
+				instance: ruler,
+			}
+		}(&rulers.Instances[i])
+	}
 
-	// Concurrently fetch rules from all rulers. Since rules are not replicated,
-	// we need all requests to succeed.
-	jobs := concurrency.CreateJobsFromStrings(rulers.GetAddresses())
-	err = concurrency.ForEach(ctx, jobs, len(jobs), func(ctx context.Context, job interface{}) error {
-		addr := job.(string)
+	groupCounterMap := make(map[string]*[]*groupCounter)
 
-		rulerClient, err := r.clientsPool.GetClientFor(addr)
-		if err != nil {
-			return errors.Wrapf(err, "unable to get client for ruler %s", addr)
+	enough := false
+	quorum := (rulerRing.ReplicationFactor() / 2) + 1
+	numSucceeded := 0
+	minSucceededForQuorum := len(rulers.Instances) - quorum + 1
```

Comment:
Could we explain a little bit for how this calculation done?

Reply:
I've added some comments to these lines in the latest commit.

Specifically for quorum, minSucceededForQuorum, and maxErrors when quorum=strong:

quorum
- quorum means that a quorum-accepted rulegroup version is returned by more than half of the replicas in the replication set.  For RF=3, this would mean that 2 rulers would need to return the same rulegroup for it to be accepted by quorum.

minSucceededForQuorum
- This is the minimum number of rulers that would need to return successfully to allow us to determine a quorum (if possible).  For a tenant ring of 6 rulers, and RF=3, if 5 rulers return successfully, we can skip waiting for the 6th ruler, and proceed to merging results.

maxErrors
- Maximum number of error results we can tolerate.  Once this is reached, the method will return an error.  For RF=3, this would be 2.  If 2 rulers are not responding, those 2 rulers could have rulegroups that would have formed a quorum if they had been able to return.